data-science-onramp/data-cleaning/clean.py

```diff
@@ -0,0 +1,179 @@
+    if start and end and not duration:
+        duration = end - start
+    elif duration and end and not start:
+        start = end - duration
+    elif duration and start and not end:
+        end = start + duration
+
+    # Convert from datetime/timedelta
+    if duration:
+        duration = int(duration.total_seconds())
+    if start:
+        start = start.strftime(time_format)
+    if end:
+        end = end.strftime(time_format)
+
+    return (duration, start, end)
+        
+def compute_duration(duration, start, end):
+    '''Calculates duration from start and end time if null.'''
+    return compute_time(duration, start, end)[0]
+
+def compute_start(duration, start, end):
+    '''Calculates start time from duration and end time if null.'''
+    return compute_time(duration, start, end)[1]
+
+def compute_end(duration, start, end):
+    '''Calculates end time from duration and start time if null.'''
+    return compute_time(duration, start, end)[2]
+
+def backfill(df):
+    path = 'gs://' + BUCKET_NAME + '/clean_citibike_data' + '.csv.gz'
+    df.write.options(codec='org.apache.hadoop.io.compress.GzipCodec').csv(path)
+
+def main():
+    # Create a SparkSession under the name 'clean'. Viewable via the Spark UI
+    spark = SparkSession.builder.appName('clean').getOrCreate()
+
+    # Whether to backfill data to GCS bucket
+    upload = True
+
+    # Check if table exists
+    try:
+        df = spark.read.format('bigquery').option('table', TABLE).load()
+    except Py4JJavaError:
+        print(f'{TABLE} does not exist. ')
+        return
+
+    # Single-parameter column transformations
+    udfs = {
+        'start_station_name': (station_name, StringType()),
+        'end_station_name': (station_name, StringType()),
+        'tripduration': (trip_duration, IntegerType()),
+        'usertype': (user_type, StringType()),
+        'gender': (gender, StringType()),
+        'start_station_latitude': (angle, FloatType()),
+        'start_station_longitude': (angle, FloatType()),
+        'end_station_latitude': (angle, FloatType()),
+        'end_station_longitude': (angle, FloatType())
+    }
+
+    for name, udf in udfs.items():
+        df = df.withColumn(name, UserDefinedFunction(*udf)(name))
+
+    # Multiple-parameter column transformations
+    multi_udfs = {
+        'tripduration': {
+            'udf': (compute_duration, IntegerType()),
+            'params': ('tripduration', 'starttime', 'stoptime')
+        },
+        'starttime': {
+            'udf': (compute_start, StringType()),
+            'params': ('tripduration', 'starttime', 'stoptime')
+        },
+        'stoptime': {
+            'udf': (compute_end, StringType()),
+            'params': ('tripduration', 'starttime', 'stoptime')
+        }
+    }
+
+    for name, obj in multi_udfs.items():
+        df = df.withColumn(name, UserDefinedFunction(*obj['udf'])(*obj['params']))
+
+    if upload:
+        backfill(df)
+    else:
+        df.show(n=100)
```

Comment:
I think you should get a sample of the dataset and then show it like this: `new_df.sample(False, 0.0001).show(n=100)` This is because when I checked the job output, all of the genders were female for some reason? Not showing a random sample might cause problems when testing.